### ISSUE TITLE
Card details not complete when using confirmPayment

### ISSUE THREAD

**Issue by hierror** [stripe/stripe-react-native#508]
**Describe the bug**
After navigating through my app and entering the payment screen, my payments always fails with the message `Card details not complete`, which is exactly the same error that these other issues were reporting: #350, #371 and #502. This error was supposed to be fixed after the [release of v0.2.0](https://github.com/stripe/stripe-react-native/issues/391#issuecomment-884033869), but it is the same for me even after upgrading the library. The error occured to every device that I have used for tests, both Android and iOS.

**To Reproduce**
Follow the steps: [Accepting a Payment](https://stripe.com/docs/payments/accept-a-payment).

**Expected behavior**
The payment should be confirmed and show up in Stripe Dashboard.

**Additional context**
This seems to be related to how the `CardField` is being deallocated by the `CardViewManager`, since applying the #379 changes fixed the error for me, but it was rejected since it might cause unexpected behavior.

**Comment by arekkubaczkowski** [comment]
@hierror as I am not able to reproduce it on my env, can you provide more information about your env details? also please provide code snippet of your implementation so that I can take a look on it closer

**Comment by arekkubaczkowski** [comment]
> This seems to be related to how the CardField is being deallocated by the CardViewManager, since applying the #379 changes fixed the error for me, but it was rejected since it might cause unexpected behavior.

We changed the modules communication approach so it's not the case right now

**Comment by hierror** [comment]
> @hierror as I am not able to reproduce it on my env, can you provide more information about your env details? also please provide code snippet of your implementation so that I can take a look on it closer

Sure, I'm using the CardField in the same view that I am currently invoking this function below. When I navigate back and forth between the different screen of my app, I always receive `Card details not complete` when making a payment. Could it be related to the CardField unmounting?

```js
const handlePayPress = async () => {
    const { data: paymentIntent } = await getPaymentIntentClientSecret()

    const billingDetails = getBillingDetails(props)

    const params = {
      type: 'Card',
      setupFutureUsage: 'OffSession',
      billingDetails,
    }

    const { client_secret: clientSecret } = paymentIntent

    const { error: confirmPaymentError } = await confirmPayment(
      clientSecret,
      params
    )

    if (confirmPaymentError) {
      console.log(confirmPaymentError)
    }
  }
```

**Comment by fgkrieger13** [comment]
I have also been experiencing this issue with 'card details not complete' on confirmPayment(). I went so far as to remove the CardField and use 'react-native-vertical-credit-card-input' then retouch all of my cardDetails to match the format shown in Stripes docs. I'm successfully creating customers and subscriptions with the clientSecret being returned, but whenever I run the confirmPayment I'm hit wit the same error log. Just thought I would share my code snippet on here as well since I'm receiving the same error with a slightly different set up. Thanks!

Here is my CardData
==========================
`PAY Object {
  "cvc": "242",
  "email": "fakeemail@gmail.com",
  "expiryMonth": "04",
  "expiryYear": "2024",
  "last4": "4242",
  "name": "Test",
  "number": "4242424242424242",
}`

Below is the code for my submit payment button onPress.
===============================================
`  const onSubmit = async () => {
    const newCustomer = await createCustomer(user.email);

    const newSubscription = await createSubscription(
      "price_1JNOTMYREALID",
      newCustomer.customer.id
    );

    const cardDetails = {
      number: cardData.values.number.replace(/[^0-9]/g, ""),
      email: user.email,
      cvc: cardData.values.cvc,
      name: cardData.values.name,
      expiryMonth: cardData.values.expiry.substring(0, 2),
      expiryYear: `20${cardData.values.expiry.substring(3, 5)}`,
      last4: cardData.values.number.substr(cardData.values.number.length - 4),
    };

    console.log("PAY", cardDetails, "CLIENT", newSubscription);

    const { error, paymentIntent } = await confirmPayment(
      newSubscription.clientSecret,
      {
        type: "Card",
        cardDetails,
      }
    );

    console.log("WOOT", paymentIntent, error);
  };`


Error Result
============
`Object {
  "code": "Failed",
  "declineCode": null,
  "localizedMessage": "Card details not complete",
  "message": "Card details not complete",
  "stripeErrorCode": null,
  "type": null,
}`

**Comment by arekkubaczkowski** [comment]
@fgkrieger13 you can't use 3rd party components for collecting card details, `confirmPayment` and other payment methods don't accept `cardDetails` property, for PCI compliance reason library hand it over under the hood. 
if you face with `card details not completed` error using CardField or CardForm, please provide information about your env, library versions, and expo if applicable

**Comment by arekkubaczkowski** [comment]
> since applying the #379 changes fixed the error for me, but it was rejected since it might cause unexpected behavior.

@hierror please make sure that you've upgraded library properly and all of the changes are applied, after last refactor those changes shouldn't have any impact on how library works

**Comment by hierror** [comment]
> > since applying the #379 changes fixed the error for me, but it was rejected since it might cause unexpected behavior.
> 
> @hierror please make sure that you've upgraded library properly, after last refactor those changes shouldn't have any impact on how library works

I updated the library to the version v.0.2.0 and the error kept ocurring. Thus, I downgraded the version to v0.1.5 and applied the changes suggested on #379, which fixed the error. Yet, I do know that those changes could cause unexpected behavior, which makes them not really helpful. I think that we are having some trouble with the component `CardField` mounting and unmounting, since this error only happens when I navigate back and forth into the payment screen before trying to make a payment.

**Comment by arekkubaczkowski** [comment]
@hierror ok thanks, we will take a look on this again

**Comment by arekkubaczkowski** [comment]
@hierror which library for navigation do you use? I am not able to reproduce it with `react-navigation v5`. Are you able to provide a demo repository where I could reproduce your issue?

**Comment by markl-vesper** [comment]
same issue here

I've also just started process from tipsi to react-native-stripe. Huge amount of work to get all the RN packages updated long with Android build issues as result.

Gradle 6.9
Gradle Plugin 4.1.2

"@stripe/stripe-js": "1.17.0",
"@stripe/stripe-react-native": "0.2.0"

When calling 

```
                    const {paymentMethodId, error} = await createPaymentMethod({
                        type: 'Card'
                    });
```
console log reports

createPaymentMethod response:  {"stripeErrorCode":null,"message":"Card details not complete","declineCode":null,"type":null,"localizedMessage":"Card details not complete","code":"Failed"}

stripe imported via 

`import { CardField, useStripe, confirmPayment, createPaymentMethod } from "@stripe/stripe-react-native";`

Card defined as 

```
                    <View style={Styles.inputContainerStyle}>
                        <CardField
                            postalCodeEnabled={false}
                            cardStyle={{
                                backgroundColor: '#FFFFFF',
                                textColor: '#000000',
                            }}
                            style={{
                                width: '100%',
                                height: 50,
                                marginVertical: 30,
                            }}
                            onCardChange={(cardDetails) => {
                                this.handleStripeParamsChange(cardDetails.complete, cardDetails)
                            }}
                        />
                    </View>
```

**Comment by arekkubaczkowski** [comment]
@hierror @markl-vesper can you check if above workaround works for you as well?

**Comment by hierror** [comment]
> @hierror @markl-vesper can you check if above workaround works for you as well?

I tried the above workaround and many more but none of them worked. The only time I was able to get around this bug was before the release of v0.2.0, when I forked the repository and applied the #379 changes. I know we can't use this PR as the solution, since it can lead to unexpected behavior and the code is already very different after the last release. Yet I wonder what did it do that fixes this bug? 🤔

**Comment by arekkubaczkowski** [comment]
@hierror are you able to provide repository where I could reproduce this specific case?

**Comment by markl-vesper** [comment]
just been chatting with stripe devs over on discord

the only way to get working currently was to move StripeProvider to directly above the CardForm. Loading the provider in different level of heirarchy fails. He believes we should be able to load at top level of app or further up the chain and for it to cascade thru. It could come down to how your navigation is stacked etc

below works fine

  <StripeProvider ... }>
       <CardField ... />
 </StripeProvider>

**Comment by arekkubaczkowski** [comment]
@hierror if you are not able to provide such repo, we can catch up on discord and we can try together debug what is going on there

**Comment by arekkubaczkowski** [comment]
closing due to inactivity, @hierror feel free to reopen if you still having this issue

**Comment by hierror** [comment]
hey @arekkubaczkowski, i think it would be great to debug it together through discord

**Comment by KryptKode** [comment]
I am having the same issue. None of the workarounds has worked. I am using `v0.2.1`

**Comment by arekkubaczkowski** [comment]
@KryptKode please provide all of necessary information about your environment, also code snippet would be useful.

**Comment by maxgenest** [comment]
Same problem here but with `confirmSetupIntent`. `handleCardaction` works perfectly fine for a simple payment but to deal with setup intents I need to use confirmSetupIntent.

With `confirmSetupIntent` I have a "card details not complete" error on the first try to pay with a card with 3D secure (I use 4000 0000 0000 3220). 
On a second try with the same or another card (with 3D secure) it works fine. 

My environment : 
```
"react-native": "0.64.2"
"@stripe/stripe-react-native":  "0.2.1"
```

**Comment by arekkubaczkowski** [comment]
@maxgenest 
> please provide all of necessary information about your environment, also code snippet would be useful.

^

**Comment by maxgenest** [comment]
> @maxgenest
> 
> > please provide all of necessary information about your environment, also code snippet would be useful.
> 
> ^

Gradle: 6.7
Gradle Plugin: 4.1.0

"@stripe/stripe-react-native": "0.2.1"
"react-navigation": "4.4.4",

Stripe imported with `<StripeProvider>` (also tried with initStripe without success)

My Card component : 
```
<CardField
            postalCodeEnabled={false}
            placeholder={{
              number: '4242 4242 4242 4242',
            }}
            style={styles.cardFieldStyle}
            onCardChange={onCardChange}
          />
```
````
const {paymentMethod, error} = await createPaymentMethod({
      type: 'Card',
      billingDetails,
    });
````
My setupIntent confirmation method : 
```
const {error} = await confirmSetupIntent(clientSecret, {
        type: 'Card',
      });
      if (error) {
        Alert.alert('an error occured');
        console.log('error', error);
      } else {
        handlePaymentSuccess(true);
      }
```

error result : `error: {stripeErrorCode: null, message: 'Card details not complete', declineCode: null, type: null, localizedMessage: 'Card details not complete', …}`

**Comment by arekkubaczkowski** [comment]
Why do you create payment method when you confirm setup intent? Please follow the docs or our example:
https://github.com/stripe/stripe-react-native/blob/master/example/src/screens/SetupFuturePaymentScreen.tsx#L62
https://stripe.com/docs/payments/save-and-reuse?platform=react-native

**Comment by arekkubaczkowski** [comment]
if you still having this problem please create another issue and use bug report template

**Comment by KryptKode** [comment]
> I am having the same issue. None of the workarounds has worked. I am using `v0.2.1`. Any other ideas on how to work around this @arekkubaczkowski ?



> @KryptKode please provide all of necessary information about your environment, also code snippet would be useful.

Thanks @arekkubaczkowski . I was able to resolve the issue. Turned out that navigation to the screen for cards was triggered from two places and when I removed one, it worked.

**Comment by hugoh59** [comment]
I'm also unable to make createPaymentMethod work. Here is my code:

```
import { initStripe, CardForm, createPaymentMethod } from '@stripe/stripe-react-native';
...

const AddCard = ({navigation}) => {

useEffect(() => {
    get('/api/v3/config/', null, global.token)
    .then(res => {
      console.log('configHere', res)
      return res.json()
    })
    .then((res) =>
      initStripe({
        publishableKey: res.data.payment_methods.filter(
          (i) => i.name === "CB - Stripe SCA (Intents)"
        )[0].public_key,
        // merchantIdentifier: 'merchant.identifier',
      })
      .then(res => {
        console.log('stripeRes', res)
      })
    )
  }, []);
const sendCardToSripe = async () => {
    console.log('cardDetails', cardDetails)
    if (isStripeSca) {
      try {
        if(!cardDetails.complete) {
          return Alert.alert(
            t(trans['error_message']),
            t(trans['error_missing_fields'])
          )
        }
        const {paymentMethod, error} = await createPaymentMethod({
          type: 'Card'
        });

        console.log('paymentMethod', paymentMethod)
        if(paymentMethod) {
          const response = await addPaymentMeanByStripePM(
            token, paymentMethod.id, dispatch, name
          )
          console.log('response', response)
          if(response.data.payment.means[0]) {
            setCard(response.data.payment.means[0])
          }
        }
}
...
return(
<CardForm
                postalCodeEnabled={false}
                autofocus={true}
                placeholder={{
                  number: '0123 4567 8912 3456',
                }}
                onFormComplete={(cardDetails) => {
                  console.log('card details', cardDetails);
                  setCardDetails(cardDetails)
                }}
/>
  <Button
            size='button_regular'
            onPress={() => {
              sendCardToSripe();
            }}
            text={t(trans['add_card.button'])}
 />
)
```

**Comment by JS-junior** [comment]
> We had the same issue and found a solution which is working for us. The strange thing was in the beginning everything went like it should an no error was thrown. At some point we were getting the `Card details not complete` even though the `CardField` onCardChange event responding with `complete: true`. 🤯
> 
> To fix this we needed to make sure that the `CardField` component has rendered when Stripe is initializing (If it is rendered conditionally, e.g. after some loading has finished ... it wont work). To achieve this we ditched the `StripeProvider`, switched to `initStripe()` and placed it in the same component in which `CardField` is used. Should look something like this:
> 
> ```ts
> const Payment = () => {
>   // Init Stripe
>   useEffect(() => {
>     async function initialize() {
>       await initStripe({
>         publishableKey: PUBLIC_KEY,
>       });
>     }
>     initialize().catch(console.error);
>   }, []);
> 
>   return (
>     <View>
>       <CardField
>         postalCodeEnabled={false}
>         onCardChange={(cardDetails: CardFieldInput.Details) => {
>           setCardDataComplete(cardDetails.complete);
>         }}
>      />
>    </View>
>   );
> }
> ```
> 
> And voila it's working again!! 🥳
> 
> Edit: Almost forgot the mention the versions we use
> 
> ```
> "expo": "^42.0.0",
> "@react-navigation/native": "^5.9.4",
> "@stripe/stripe-react-native": "0.1.4",
> ```

It worked!!
🔥🔥🔥🔥🔥
Thanks,
It's been  a week, I have been struggling with it.
Please include this is in stripe docs and npm

**Comment by arekkubaczkowski** [comment]
I am closing this due to inactivity.
@hierror  I am aware about your issue and I believe it's strongly related with your custom navigation solution.

There is one known issue which might be related to this https://github.com/stripe/stripe-react-native/issues/625
please get familiar with this thread and make sure that you are not rendering CardField within FlatList

**Comment by Emehado** [comment]
I had the same issue but mine was due to rendering. I set a loading state to true and that replaces the entire page which includes the CardField element with a spinner when I am getting the client secret from the server, and the loading state remains true until confirmPayment has been resolved. I think confirmPayment needs the CardField Element to be rendered whiles executing. I hope this helps someone.

**Comment by alexsegura** [comment]
This is still an issue. On my side, I have noticed that it starts triggering the `Card details not complete` error **after a React refresh** while developing. To test cards without this annoying error, I have to go back one screen (I use React Navigation 5.x), and go forward to my payment screen, so that everything is "reloaded" properly. 

So it seems to be related with the mounting of what passes the card details "under the hood".

**Comment by NamNP06** [comment]
Hello everybody . I faced problem the same issue. when I navigate screen : I received the same error: Card detail not complete. I try to fix problem but not success. Please help me or you can check code library. sorry If i wrong write. Some day, I will release components stripe for customer. 
I use workaround and code  the same @ingowonner:
    "react-router-dom": "^5.2.0",
    "react-native": "0.64.1",
I am not use StripeProvider and swtich to UseEffect with initStripe 
 useEffect(() => {
    async function initialize() {
      await initStripe({
        publishableKey: PUBLIC_KEY,
      });
    }
    initialize().catch(console.error);
  }, []);

  return (
    <View>
      <CardField
        postalCodeEnabled={false}
        onCardChange={(cardDetails: CardFieldInput.Details) => {
          setCardDataComplete(cardDetails.complete);
        }}
     />
   </View>

Please help me !. Thank you !

**Comment by NorthBlue333** [comment]
Hi @alexsegura
Happens to me too, using Expo SDK 44 and Stripe SDK 0.2.3.

**Comment by Roundups** [comment]
I am experiencing this issue with Expo SDK44 and Stripe SDK 0.2.3. However it does work in iOS, only returns this error response on android. This is with `confirmSetupIntent` instead though.

iOS does trigger the secure checkout window but it errors on android before this happens so maybe the issue lies there? Any advice is appreciated.

**Comment by SashaMihalache** [comment]
Happened to me too on Android only.

The issue was because I was conditionally rendering the Credit Card form only after some loading logic was done.

As soon as I removed the condition and let the form render when the component mounted it worked alright.

**Comment by BerkeAras** [comment]
Same issue — none of the Workarounds work :(

**Comment by gui10l1** [comment]
Same issue here. None of the solutions worked for me...

**Comment by raajnadar** [comment]
@SashaMihalache you are right if we remove the condition it works fine! Not sure how

**Comment by hislordshipprof** [comment]
have been facing issues with setup intent always giving me no such setup intent 
this is my code base , 
`const { confirmSetupIntent, loading } = useConfirmSetupIntent();

const handlePayPress = async () => {
        if(!cardDetails?.complete || !email ){
          Alert.alert("Please fill all the fields and Email")
          return
        }
        const billingDetails = {   email: email };
try {
  const { setupIntent, error } = await confirmSetupIntent(clientSecret, {
    type: 'Card',
    paymentMethodId: "",
      paymentMethodData: {
      billingDetails: billingDetails,
      }
     
  });

  if (error) {
    console.log("my error",error)
  }
  else {
    console.log("===my setupIntent",setupIntent)
  }

} catch (err) {
  console.log("-----errr",err)
}
      
      };`

**Comment by orlandozeta** [comment]
@SashaMihalache you are right works fine removing the conditional!

**Comment by didiabel** [comment]
i still have this issue, even if i initialize stripe in a useefect on the cardfield component, the error still persists...

**Comment by usamamoinakhter** [comment]
Had been facing this issue for a while, but after hours of debugging finally figured out what the problem was. The problem is that I had multiple screens where cardfield was being rendered lets call them **screenA** and **screenB** for clarity's sake, when I want to go to **screenB** I need to make sure that **screenA** isn't in the navigation stack and vice versa. Whenever there are multiple screens with cardfield in the stack navigator it causes this issue for me.

**Comment by edi** [comment]
I've been greeted with the same error message. It only seems to happen when I use this card `4000 0027 6000 3184`. Any other card seems to work just fine. It works when I call `confirmSetupIntent` using a `CardField` (3DS goes through just fine), but when I try confirm an actual payment intent which comes back as "requires_action", the error message pops up every single time, `Card details not complete`, even though the same exact flow with any other card works.

So I tend to think it's something on Stripe's end with that card.

**Comment by jsam6** [comment]
Has this issue been resolved?
I have tried all suggestions above and yet to find a solution.

Initiating stripe from useEffect and removing all stacks of CardField does not solve my issue.

**Comment by edi** [comment]
I have figured out the issue after some due dilligence research.

Basically, in v0.24.0 a breaking change happened with the `confirmPayment` method, which now requires different parameters, as explained here: https://github.com/stripe/stripe-react-native/commit/99bcc2722d8234e991529e4fbc867e220e90f863

`type` changed to `paymentMethodType`
`params` changed to `paymentMethodData`

So a valid `confirmPayment` call would now look like this:

```
const {paymentIntent, error} = await confirmPayment(intent.secret, {
    paymentMethodType: 'Card',
    paymentMethodData: {paymentMethodId: intent.method}
})
```

I think that in the end, the library should have some required props, given we are not within any stable release yet. 
The root of the issue above, was that since it couldn't find the `paymentMethodId` as we were passing the wrong parameters, it would try to fetch data from the `CardField` component, which is most likely not mounted, as you're already trying to confirm a previously linked `paymentMethod`, hence, the `Card details not complete` error.

Regards

**Comment by GeuntaBuwono** [comment]
if you guys still stuck on `Card details not complete` 
component should not be re-render, so if you have loading state need to take out, createPaymentMethod expect to not close it until createToken is done. 

and for `CardField` component need to add default countryCode `countryCode="AU"` for example 

i've been using `"@stripe/stripe-react-native": "0.19.0",`

**Comment by kidmysoul** [comment]
while my problem is because i saved clientSecret in redux store when get from createPaymentIntend and 

**this triggered a re-render!**! so when i used it for confirmPayment. error happened.

**Comment by CTB333** [comment]
I am not sure if this is helpful, but to anyone that comes upon this with possibly the same problem maybe it will. The cause of this error for me was writing in a conditional color change into the style of the card field, moving the condition outside of the card field fixed the error message.

**Comment by arun057** [comment]
To everyone who is stuck with this issue, DO NOT dispose (close) of the card field until the payment is confirmed.

**Comment by kharitonovAL** [comment]
> We had the same issue and found a solution which is working for us. The strange thing was in the beginning everything went like it should an no error was thrown. At some point we were getting the `Card details not complete` even though the `CardField` onCardChange event responding with `complete: true`. 🤯
> 
> To fix this we needed to make sure that the `CardField` component has rendered when Stripe is initializing (If it is rendered conditionally, e.g. after some loading has finished ... it wont work). To achieve this we ditched the `StripeProvider`, switched to `initStripe()` and placed it in the same component in which `CardField` is used. Should look something like this:
> 
> ```ts
> const Payment = () => {
>   // Init Stripe
>   useEffect(() => {
>     async function initialize() {
>       await initStripe({
>         publishableKey: PUBLIC_KEY,
>       });
>     }
>     initialize().catch(console.error);
>   }, []);
> 
>   return (
>     <View>
>       <CardField
>         postalCodeEnabled={false}
>         onCardChange={(cardDetails: CardFieldInput.Details) => {
>           setCardDataComplete(cardDetails.complete);
>         }}
>      />
>    </View>
>   );
> }
> ```
> 
> And voila it's working again!! 🥳
> 
> Edit: Almost forgot the mention the versions we use
> 
> ```
> "expo": "^42.0.0",
> "@react-navigation/native": "^5.9.4",
> "@stripe/stripe-react-native": "0.1.4",
> ```

Man, you are awesome! Im not from react, but from flutter, and this is worked same way:

1. init Stripe
2. fill card details in CardField
3. create token

thanks!

**Comment by umairm1alik** [comment]
I am using redux toolkit and redux persist in my project, previously I was wrapping like this and it was not working.
<StripeProvider publishableKey="">
      <Provider store={store}>
        <PersistGate
          loading={<MainLoader />}
          persistor={persistor}
          onBeforeLift={() =>
            new Promise(resolve => setTimeout(resolve, 3000))
          }>
            <Navigation />
        </PersistGate>
      </Provider>
    </StripeProvider>

and then I just move stripe provider within Provider now its working .
<Provider store={store}>
        <PersistGate
          loading={<MainLoader />}
          persistor={persistor}
          onBeforeLift={() =>
            new Promise(resolve => setTimeout(resolve, 3000))
          }>
          <StripeProvider publishableKey="">
            <Navigation />
           </StripeProvider>
        </PersistGate>
      </Provider>
